FprimeZephyrReference/Components/Authenticate/Authenticate.cpp

```diff
@@ -0,0 +1,532 @@
+    // decrement the size of the data to remove the footer
+    data.setSize(data.getSize() - SECURITY_TRAILER_LENGTH);
+
+    // the first two bytes are the SPI
+    U32 spi = (static_cast<U32>(securityHeader[0]) << 8) | static_cast<U32>(securityHeader[1]);
+
+    // the next four bytes are the sequence number
+    U32 sequenceNumber = (static_cast<U32>(securityHeader[2]) << 24) | (static_cast<U32>(securityHeader[3]) << 16) |
+                         (static_cast<U32>(securityHeader[4]) << 8) | static_cast<U32>(securityHeader[5]);
+
+    const AuthenticationConfig authConfig = this->lookupAuthenticationConfig(spi);
+    const std::string type_authn = authConfig.type;
+    const std::string& key_authn = authConfig.key;
+
+    // check the sequence number is valid
+    // cast trailer to U32
+    bool sequenceNumberValid = this->validateSequenceNumber(sequenceNumber, this->get_SequenceNumber());
+    if (!sequenceNumberValid) {
+        U32 newCount = this->rejectedPacketsCount.load() + 1;
+        this->rejectedPacketsCount.store(newCount);
+        this->tlmWrite_RejectedPacketsCount(newCount);
+        this->persistToFile(REJECTED_PACKETS_COUNT_PATH, newCount);
+        contextOut.set_authenticated(0);
+        this->dataOut_out(0, data, contextOut);
+        return;
+    } else {
+        // increment the stored sequence number
+        U32 newSequenceNumber = sequenceNumber + 1;
+        this->sequenceNumber.store(newSequenceNumber);
+        this->tlmWrite_CurrentSequenceNumber(newSequenceNumber);
+        this->persistToFile(SEQUENCE_NUMBER_PATH, newSequenceNumber);
+    }
+
+    Fw::Buffer computedHmac = this->computeHMAC(securityHeader, 6, data.getData(), data.getSize(), key_authn);
+    const U8* computedHmacData = computedHmac.getData();
+    const FwSizeType computedHmacLength = computedHmac.getSize();
+
+    // Clean up the allocated memory when done (Fw::Buffer doesn't manage memory)
+    // Store non-const pointer for deletion (getData() returns const, but we allocated with new[])
+    U8* hmacDataToDelete = const_cast<U8*>(computedHmacData);
+
+    if (computedHmacData == nullptr || computedHmacLength < 8) {
+        if (hmacDataToDelete != nullptr) {
+            delete[] hmacDataToDelete;
+        }
+        this->log_WARNING_HI_InvalidHash(contextOut.get_apid(), spi, sequenceNumber);
+        U32 newCount = this->rejectedPacketsCount.load() + 1;
+        this->rejectedPacketsCount.store(newCount);
+        this->tlmWrite_RejectedPacketsCount(newCount);
+        this->persistToFile(REJECTED_PACKETS_COUNT_PATH, newCount);
+        contextOut.set_authenticated(0);
+        this->dataOut_out(0, data, contextOut);
+        return;
+    }
+
+    bool hmacValid = this->compareHMAC(computedHmacData, securityTrailer, computedHmacLength);
+
+    // Memory is no longer needed after comparison - delete before any return
+    delete[] hmacDataToDelete;
```

Comment:
I don't know much about cpp memory management but my gut tells me that we could increase safety by moving this to a separate fn where the scope of `hmacDataToDelete` only exists in that one function and cannot be accidentally referenced later in this function without throwing a compiler error.

Also noting that in golang we have a nice `defer` fn which, if we had a similar construct here, would allow us to call the deletion immediately after creation with a deferral so that it happens only after the scope has been closed. Someone may have made a similar construct if you want to try it: https://www.rioki.org/2024/03/24/cpp-defer.html